lib/api/core/indexCache.js

```diff
@@ -32,113 +33,135 @@ const
+
+    this.internalStorage = this.kuzzle.services.internalStorage;
+    this.publicStorage = this.kuzzle.services.publicStorage;
+
+    // Structure:
+    //   myIndex: {
+    //     indexType: 'public',
+    //     collections: []
+    //   }
+    this.indexes = {};
   }
 
+  /**
+   * Initializes the index cache with existing indexes and collections
+   *
+   * @returns {Promise}
+   */
   init () {
-    const internalEngine = this.kuzzle.internalEngine;
-
-    return internalEngine.listIndexes()
-      .then(indexes => indexes.filter(index => index !== this.kuzzle.internalEngine.index))
-      .then(indexes => {
-        const promises = [];
-
-        for (const index of indexes) {
-          this.indexes[index] = [];
-          this.defaultMappings[index] = _.cloneDeep(this.commonMapping);
-
-          promises.push(
-            internalEngine.listCollections(index)
-              .then(collections => {
-                const applyMappingPromises = [];
-
-                for (const collection of collections) {
-                  this.indexes[index].push(collection);
-
-                  applyMappingPromises.push(
-                    this.kuzzle.internalEngine.applyDefaultMapping(index, collection, this.defaultMappings[index])
-                  );
-                }
-
-                return Promise.all(applyMappingPromises);
-              })
-              .then(defaultMappings => {
-                // We are working on the same index so the returned mappings are all the same
-                this.defaultMappings[index] = defaultMappings[0];
-              })
-          );
-        }
+    const promises = [];
 
-        return Bluebird.all(promises);
-      })
-      .then(() => internalEngine.listAliases())
-      .then(aliases => {
-        for (const alias of Object.keys(aliases)) {
-          const original = aliases[alias];
-          this.indexes[alias] = this.indexes[original];
-          this.defaultMappings[alias] = this.defaultMappings[original];
-        }
-      });
+    promises.push(
+      this.internalStorage.listIndexes()
+        .then(indexes => this._addCollections(indexes, 'internal')));
+
+    promises.push(
+      this.publicStorage.listIndexes()
+        .then(indexes => this._addCollections(indexes, 'public')));
+
+    return Bluebird.all(promises);
   }
 
-  add (index, collection, notify = true) {
+  /**
+   * Lists collections from all indexes passed in argument and add them to the cache
+   *
+   * @param {Array.<String>} indexes - Index names
+   * @param {String} indexType - Index type (internal or public)
+   *
+   * @returns {Promise}
+   */
+  _addCollections (indexes, indexType) {
+    const promises = [];
+
+    for (const index of indexes) {
+      this.indexes[index] = {
+        indexType,
+        collections: []
+      };
+
+      promises.push(
+        this.internalStorage.listCollections(index)
+          .then(collections => {
+            for (const collection in collections) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Restrict what this loop acts on by testing each property. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AForIn)